File: week5/python/844. Backspace String Compare.py
class Solution:
    def backspaceCompare(self, S: str, T: str) -> bool:
        return self.getFinalString(S) == self.getFinalString(T)
    
    def getFinalString(self, S):
        result = ''
        bs_count = 0
        for i in range(len(S)-1, -1, -1):
            if S[i] == '#':
                bs_count += 1
            else:
                if bs_count > 0:
                    bs_count -= 1
                else:
                    result += S[i]
        return result

File: week5/python/844. Backspace String Compare.java
class Solution {
    public boolean backspaceCompare(String S, String T) {
        return delete(S).equals(delete(T));
    }
    
    private String delete(String s) {
        Stack<Character> st = new Stack<>();
        char[] ch = s.toCharArray();
        for (int i = 0 ; i < s.length(); i++) {
            if (ch[i]!='#') {
                st.push(ch[i]);
            } else {
                if (!st.isEmpty()) {
                    st.pop();
                }
            }
        }
        
        return String.valueOf(st);
    }
}
